cscs-checks/tools/profiling_and_debugging/perftools.py

```diff
@@ -22,48 +22,51 @@ def __init__(self, lang, **kwargs):
 
+        # NOTE: Reduce time limit because for PrgEnv-pgi even if the output
+        # is correct, the batch job uses all the time.
+        self.time_limit = (0, 1, 0)
+
         self.num_gpus_per_node = 1
         self.executable = 'perftools_check'
-
         self.prgenv_flags = {
-            'PrgEnv-cray':  ' -h nomessage=3140 -homp ',
-            'PrgEnv-gnu':   ' -fopenmp ',
-            'PrgEnv-intel': ' -openmp ',
-            'PrgEnv-pgi':   ' -mp ',
+            'PrgEnv-cray': ['-h nomessage=3140', '-homp'],
+            'PrgEnv-gnu': ['-fopenmp'],
+            'PrgEnv-intel': ['-openmp'],
+            'PrgEnv-pgi': ['-mp']
         }
-
         if self.current_system.name == 'kesch':
             # `-lcudart -lm` must be passed explicitly on kesch
-            self.prgenv_flags['PrgEnv-gnu'] = ' -fopenmp -lcudart -lm '
+            self.prgenv_flags['PrgEnv-gnu'] = ['-fopenmp', '-lcudart', '-lm']
 
         self.sanity_patterns = sn.assert_found('Table 1:  Profile by Function',
                                                self.stdout)
-        self.modules  = ['perftools-base', 'perftools-lite', 'cudatoolkit']
-        self.makefile = 'Makefile_perftools'
-        self.sourcesdir = os.path.join('src', lang)
-        self.maintainers = ['MK', 'JG']
-        self.tags = {'production'}
+        self.modules = ['perftools-lite', 'craype-accel-nvidia60']
+        self.build_system = 'Make'
+        self.build_system.makefile = 'Makefile_perftools'
+        if lang == 'Cpp':
+            self.sourcesdir = os.path.join('src', 'C++')
+        else:
+            self.sourcesdir = os.path.join('src', lang)
+
+        # NOTE: Restrict concurrency to allow creation of Fortran modules
+        if lang == 'F90':
+            self.build_system.max_concurrency = 1
 
         if lang != 'Cuda':
             self.num_tasks_per_node = 2
             self.variables = {'OMP_NUM_THREADS': '2'}
         else:
             self.num_tasks_per_node = 1
 
-    def compile(self):
-        self.flags = ' -g -D_CSCS_ITMAX=1 -DUSE_MPI '
-        prgenv_flags = self.prgenv_flags[self.current_environ.name]
-        self.current_environ.cflags    = self.flags + prgenv_flags
-        self.current_environ.cxxflags  = self.flags + prgenv_flags
-        self.current_environ.fflags    = self.flags + prgenv_flags
-        self.current_environ.ldflags   = self.flags + prgenv_flags
-        super().compile(makefile=self.makefile,
-                        options='NVCCFLAGS="-arch=sm_60"')
-
-
-def _get_checks(**kwargs):
-    ret = []
-    for lang in ['C', 'C++', 'F90', 'Cuda']:
-        ret.append(PerftoolsCheck(lang, **kwargs))
+        self.maintainers = ['MK', 'JG']
+        self.tags = {'production'}
 
-    return ret
+    def setup(self, environ, partition, **job_opts):
+        super().setup(environ, partition, **job_opts)
+        flags = self.prgenv_flags[self.current_environ.name]
+        flags += ['-g', '-D_CSCS_ITMAX=1', '-DUSE_MPI']
```

Comment:
I would use the `build_system.cppflags` for the `-D` definitions and as for the `-g`, I would put it in each entry in the `prgenv_flags` dictionary. This way, you will have all the compilation flags in a single place in the `__init__()`.